provider/balance_checker.go

```diff
@@ -2,153 +2,316 @@ package provider
+	bqc     btypes.QueryClient
+	aqc     aclient.QueryClient
+	leases  map[mtypes.LeaseID]*leaseState
+	cfg     BalanceCheckerConfig
 }
 
-type BalanceCheckerConfig struct {
-	PollingPeriod           time.Duration
-	MinimumBalanceThreshold uint64
-	WithdrawalPeriod        time.Duration
+type leaseCheckResponse struct {
+	lid        mtypes.LeaseID
+	checkAfter time.Duration
+	state      respState
+	err        error
 }
 
 func newBalanceChecker(ctx context.Context,
-	bankQueryClient bankTypes.QueryClient,
+	bqc btypes.QueryClient,
+	aqc client.QueryClient,
 	accAddr sdk.AccAddress,
 	clientSession session.Session,
 	bus pubsub.Bus,
-	cfg BalanceCheckerConfig) *balanceChecker {
-	bc := &balanceChecker{
+	cfg BalanceCheckerConfig) (*balanceChecker, error) {
 
+	bc := &balanceChecker{
+		ctx:     ctx,
 		session: clientSession,
 		log:     clientSession.Log().With("cmp", "balance-checker"),
-		lc:      lifecycle.New(),
 		bus:     bus,
+		lc:      lifecycle.New(),
 		ownAddr: accAddr,
-
-		bankQueryClient: bankQueryClient,
-		cfg:             cfg,
+		bqc:     bqc,
+		aqc:     aqc,
+		leases:  make(map[mtypes.LeaseID]*leaseState),
+		cfg:     cfg,
 	}
 
+	startCh := make(chan error, 1)
 	go bc.lc.WatchContext(ctx)
-	go bc.run()
+	go bc.run(startCh)
+
+	select {
+	case err := <-startCh:
+		if err != nil {
+			return nil, err
+		}
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
 
-	return bc
+	return bc, nil
 }
-func (bc *balanceChecker) doCheck(ctx context.Context) (bool, error) {
 
-	// Get the current wallet balance
-	query := bankTypes.NewQueryBalanceRequest(bc.ownAddr, "uakt")
-	result, err := bc.bankQueryClient.Balance(ctx, query)
-	if err != nil {
-		return false, err
+func (bc *balanceChecker) runEscrowCheck(ctx context.Context, lid mtypes.LeaseID, scheduledWithdraw bool, res chan<- leaseCheckResponse) {
+	go func() {
+		select {
+		case <-bc.lc.Done():
+		case res <- bc.doEscrowCheck(ctx, lid, scheduledWithdraw):
+		}
+	}()
+}
+
+func (bc *balanceChecker) doEscrowCheck(ctx context.Context, lid mtypes.LeaseID, scheduledWithdraw bool) leaseCheckResponse {
+	resp := leaseCheckResponse{
+		lid:   lid,
+		state: respStateNextCheck,
 	}
 
-	balance := result.Balance.Amount
-	balanceGauge.Set(float64(balance.Uint64()))
+	if scheduledWithdraw {
+		resp.state = respStateScheduledWithdraw
+	}
 
-	if bc.cfg.MinimumBalanceThreshold == 0 {
-		return false, nil
+	var syncInfo *tmrpc.SyncInfo
+	syncInfo, resp.err = bc.session.Client().NodeSyncInfo(ctx)
+	if resp.err != nil {
+		return resp
 	}
 
-	tooLow := sdk.NewIntFromUint64(bc.cfg.MinimumBalanceThreshold).GT(balance)
+	if syncInfo.CatchingUp {
+		resp.err = aclient.ErrNodeNotSynced
+		return resp
+	}
 
-	return tooLow, nil
+	var dResp *dtypes.QueryDeploymentResponse
+	var lResp *mtypes.QueryLeasesResponse
+
+	// Fetch the balance of the escrow account
+	dResp, resp.err = bc.aqc.Deployment(ctx, &dtypes.QueryDeploymentRequest{
+		ID: lid.DeploymentID(),
+	})
+
+	if resp.err != nil {
+		return resp
+	}
```

Comment:
After this check if the deployment is closed, and if it is just return? We're done at that point.  It could happen if the deployment is closed as this code goroutine is running.